### ISSUE TITLE
EC2Cluster leaks VMs when adaptive scaling enabled if worker creation raises an exception

### ISSUE THREAD

**Issue by martinedgefocus** [dask/dask-cloudprovider#365]
From what I've seen from debugging the issue is this:

create_vm calls client.run_instances()
subsequent call to client.create_tags() fails, exception is not caught. 
The instance now exists, but is not returned. It will show up in the "observed" pool, but is not counted for scaling, so over time, hundreds of additional instances can be allocated, burning a huge AWS bill.

Proposed solution:
1. Retry the create_tag calls with a timeout loop similar to ip_address_key code
2. If you still fail, attempt to release the resource.

FWIW, it's also unclear to me what happens if the timeout expires to get ip_address_key at the end of the function?

I filed a related bug against the distributed code [here](https://github.com/dask/distributed/issues/6801), seems like this should be addressed in both places?

**Comment by jacobtomlinson** [comment]
Thanks for raising this, we definitely don't want to be leaking VMs and not cleaning them up.

The proposed solution sounds great, do you have any interest in opening a PR to fix this?

**Comment by martinedgefocus** [comment]
I don't think I have permissions to create a branch, and unclear how to create a PR without that?
I'll put the flat patch below to look at, but if you can give some guidance on how to create a PR, I'll do it properly.

1. Pulled out the postprocessing into configure_vm, and wrap in try/catch
2. Put create_tags in a timeout loop.
3. moved initialization of the backoff variable outside of the timeout loop for ip_address_key, else it seems pointless?
4. moved the sleep call to end of loop for ip_address_key - sleeping before attempt seems wrong?
5. changed the ip_address_key check to self.instance.get(ip_address_key) - simpler, more readable, and equivalent?
6. botocore.config.Config(retries = dict(max_attempts=10)) - without this we get errors, seems to be a fairly standard AWS problem.

```
diff --git a/dask_cloudprovider/aws/ec2.py b/dask_cloudprovider/aws/ec2.py
index 59b5cb9..97d6d91 100644
--- a/dask_cloudprovider/aws/ec2.py
+++ b/dask_cloudprovider/aws/ec2.py
@@ -20,6 +20,7 @@ from dask_cloudprovider.utils.timeout import Timeout
 try:
     from aiobotocore.session import get_session
     import botocore.exceptions
+    import botocore.config
 except ImportError as e:
     msg = (
         "Dask Cloud Provider AWS requirements are not installed.\n\n"
@@ -86,8 +87,9 @@ class EC2Instance(VMInterface):
         """
         # TODO Enable Spot support
 
+        boto_config = botocore.config.Config(retries = dict(max_attempts=10))
         async with self.cluster.boto_session.create_client(
-            "ec2", region_name=self.region
+            "ec2", region_name=self.region, config=boto_config
         ) as client:
             self.vpc = self.vpc or await get_default_vpc(client)
             self.subnet_id = (
@@ -159,46 +161,64 @@ class EC2Instance(VMInterface):
 
             response = await client.run_instances(**vm_kwargs)
             [self.instance] = response["Instances"]
-            await client.create_tags(
-                Resources=[self.instance["InstanceId"]],
-                Tags=[
-                    {"Key": "Name", "Value": self.name},
-                    {"Key": "Dask Cluster", "Value": self.cluster.uuid},
-                ],
-            )
-            self.cluster._log(
-                f"Created instance {self.instance['InstanceId']} as {self.name}"
-            )
 
-            address_type = "Private" if self.use_private_ip else "Public"
-            ip_address_key = f"{address_type}IpAddress"
+            try:    # Ensure we tear down any resources we allocated if something goes wrong
+                return await self.configure_vm(client)
+            except:
+                self.cluster._log(f"reclaiming vm because configure_vm failed {self.name}")
+                await self.destroy_vm()
+                raise
+
+
+    async def configure_vm(self, client):
+        timeout = Timeout(300, f"Failed to add tags for {self.instance['InstanceId']}")
+        backoff = 0.1
+        while timeout.run():
+            try:
+                await client.create_tags(
+                    Resources=[self.instance["InstanceId"]],
+                    Tags=[
+                        {"Key": "Name", "Value": self.name},
+                        {"Key": "Dask Cluster", "Value": self.cluster.uuid},
+                    ],
+                )
+                break
+            except Exception as e:
+                timeout.set_exception(e)
+
+            await asyncio.sleep(min(backoff, 10) + backoff % 1)
+            # Exponential backoff with a cap of 10 seconds and some jitter
+            backoff = backoff * 2
+
+        self.cluster._log(
+            f"Created instance {self.instance['InstanceId']} as {self.name}"
+        )
+
+        address_type = "Private" if self.use_private_ip else "Public"
+        ip_address_key = f"{address_type}IpAddress"
+
+        default_error = f"Failed {address_type} IP for instance {self.instance['InstanceId']}"
+        timeout = Timeout(300, default_error)
+        backoff = 0.1
+        while self.instance.get(ip_address_key) is None and timeout.run():
+            try:
+                response = await client.describe_instances(
+                    InstanceIds=[self.instance["InstanceId"]], DryRun=False
+                )
+                [reservation] = response["Reservations"]
+                [self.instance] = reservation["Instances"]
+            except botocore.exceptions.ClientError as e:
+                timeout.set_exception(e)
+            await asyncio.sleep(min(backoff, 10) + backoff % 1)
+            # Exponential backoff with a cap of 10 seconds and some jitter
+            backoff = backoff * 2
+        return self.instance[ip_address_key]
 
-            timeout = Timeout(
-                300,
-                f"Failed {address_type} IP for instance {self.instance['InstanceId']}",
-            )
-            while (
-                ip_address_key not in self.instance
-                or self.instance[ip_address_key] is None
-            ) and timeout.run():
-                backoff = 0.1
-                await asyncio.sleep(
-                    min(backoff, 10) + backoff % 1
-                )  # Exponential backoff with a cap of 10 seconds and some jitter
-                try:
-                    response = await client.describe_instances(
-                        InstanceIds=[self.instance["InstanceId"]], DryRun=False
-                    )
-                    [reservation] = response["Reservations"]
-                    [self.instance] = reservation["Instances"]
-                except botocore.exceptions.ClientError as e:
-                    timeout.set_exception(e)
-                backoff = backoff * 2
-            return self.instance[ip_address_key]
 
     async def destroy_vm(self):
+        boto_config = botocore.config.Config(retries = dict(max_attempts=10))
         async with self.cluster.boto_session.create_client(
-            "ec2", region_name=self.region
+            "ec2", region_name=self.region, config=boto_config
         ) as client:
             await client.terminate_instances(
                 InstanceIds=[self.instance["InstanceId"]], DryRun=False

```

**Comment by martinedgefocus** [comment]
https://github.com/dask/dask-cloudprovider/pull/367
I think this is correct, but not my usual workflow; apologies if it's messed up

**Comment by jacobtomlinson** [comment]
Fixed by #367
